### ISSUE TITLE
copy a yaml file in vscode results in `The class name "" is already used by another model class`

### ISSUE THREAD

**Issue by vanlooverenkoen** [serverpod/serverpod#3048]
**Describe the bug**

**To Reproduce**
1. Create a app.spy.yaml
2. Create your class
3. Copy your app.spy.yaml (cmd+c)
4. Paste your app.spy.yaml (cmd+v) in the same directory
5. app.spy copy.yaml is created
6. Update the name of the file & the name of your class
7. `The class name "App" is already used by another model class` is visible in the `problems` tab

**Expected behavior**
The copied file should not give an error because it does not exist anymore

**Screenshots**
<img width="789" alt="image" src="https://github.com/user-attachments/assets/831ad9ca-0d70-4c59-ab32-197f52b20bf4" />

**versions**
 - _VS Code plugin verion:_ v1.2.4

**Platform information**
On which platforms do you experience the issue. Please provide as much information here as you can. Where are you running the server, which front end device are you using?
macOS

**Additional context**

**Comment by SandPod** [comment]
After some quick testing it seems like the renaming event is not correctly processed by the language server.

When the file is renamed, the error tracked is preserved in the analyzer's state, referring to the old file (which now no longer exists).

### Suggested solution
When a file is renamed we should clear out any errors related to that file.

### Temporary workaround
Resolve any issues before renaming the file.

**Comment by vanlooverenkoen** [comment]
This is how I currently fix it usually.
- Tap on the error
- "the file does not exists" -> so I press save (cmd + s)
- I add random characters to make the model unique
- And I remove the file. 

This way my new model is still reserverd, and the `copy` file is removed. This clears the error in the problems tab

**Comment by vanlooverenkoen** [comment]
I have this issue since I started with serverpod but I thought it was just a bug in vscode because I had it for other files with other extensions as well, but after updating those plugins I don't see that anymore.

**Comment by SandPod** [comment]
Makes sense! I think this error has probably been there since we introduced the extension :)

**Comment by vanlooverenkoen** [comment]
@SandPod I would maybe like to fix this one, can you maybe point me in the right direction? Serverpod has improved my productivity so much, that I want to give something back!

**Comment by SandPod** [comment]
That would be great!

The problem is probably not inside of the vs code extention but in the Serverpod CLI implementation the model file analyzer.

The language server interacts with the analyzer here: https://github.com/serverpod/serverpod/blob/main/tools/serverpod_cli/lib/src/language_server/language_server.dart

The analyzer provides an interface to remove files it is tracking here: https://github.com/serverpod/serverpod/blob/2865fdcce43dfcdd2ef0eeb40b982608bb1033bb/tools/serverpod_cli/lib/src/analyzer/models/stateful_analyzer.dart#L72

My hunch tells me that we are either not correctly reporting renaming events from the language server so that these files can be removed.

We have great tests for the analyzer here: https://github.com/serverpod/serverpod/tree/main/tools/serverpod_cli/test/analyzer/models/stateful_analyzer


You should be able to simulate your scenario in a unit test and see if you can reproduce it. Inspect the code in the language server and perform similar actions in a unit test.

**Comment by vanlooverenkoen** [comment]
Thanks @SandPod I will take a look this evening, because (I hope so) fixing the bug will cost me less time than constantly fixing the problems tab 😂

**Comment by vanlooverenkoen** [comment]
Indeed the `onRenameRequest` was not implemented. 

But I can't seem to find a way to reproduce it in a unit test. There are only tests for the analyzer not for the actual communication with the language server

And it is not clear to me how I can debug it while it is connected to a vs code extension. Do you have any documentation on this somewhere?

**Comment by SandPod** [comment]
Great catch!

You are correct, we don't have unit tests for that environment unfortunately so you will need to test it out manually.

Before starting the extension, make sure to globally activate your version of the Serverpod CLI. The extension uses the globally activated version when running.

To start your local version of the extension(in vscode):
1. Open the `tools/serverpod_vscode_extention` directory in vscode.
2. Switch to the "Run and debug" tab and click "Run extension".

This will open a new vscode window where that extension is running, then you can open your local test project to run the extension.


And a little warning, `pub` has had an issue for a while where the global packages are sometimes not refreshed even if they are activated. If you notice that your changes are not applied, check this thread for more information: https://github.com/dart-lang/pub/issues/4295#issuecomment-2371670722

**Comment by vanlooverenkoen** [comment]
@SandPod THANKS! I will take a look this evening 🙏  (Thanks for being helping along the way!)

**Comment by SandPod** [comment]
No problem! I am happy to enable more contributors to Serverpod 🎉

**Comment by vanlooverenkoen** [comment]
Just adding this for future reference 👼 
-> https://github.com/serverpod/serverpod/discussions/2067#discussioncomment-8913116

**Comment by vanlooverenkoen** [comment]
```
Failed to resolve the Serverpod CLI executable. Please ensure the Serverpod CLI is installed and available on the PATH used by VS Code.
```

In vscode terminal I can run this just fine. is this related to the not refreshing of the activated package?

**Comment by SandPod** [comment]
Hmm, could be...

That warning is displayed when the extension is not able to execute the `serverpod version` when starting. This command is used to validate that the CLI is installed and available: https://github.com/serverpod/serverpod/blob/2865fdcce43dfcdd2ef0eeb40b982608bb1033bb/tools/serverpod_vscode_extension/src/extension.ts#L19-L22

Is the `serverpod` executable available in the PATH visible to the extension?

**Comment by vanlooverenkoen** [comment]
Yes indeed, I got to the same conclusion. But indeed it is just visible in my path. I will do some more debugging. because it seems that now my global serverpod_cli installation for other projects is also broken.

**Comment by SandPod** [comment]
Hey @vanlooverenkoen!

Did you manage to get the environment running?
If you have managed to partly solve this problem, feel free to open a draft PR to preserve that progress.

**Comment by vanlooverenkoen** [comment]
@SandPod I was not able to get this fixed. Maybe tonight I can take another look

**Comment by vanlooverenkoen** [comment]
I can't get my environment to work for debugging. I have no idea what I am doing wrong. 

It might be cool to have a documentation page on how to work on certain parts of the serverpod codebase. 

I found that you can override rename but I never got into that method. But I am not sure if it is my environment or anything else.

**Comment by SandPod** [comment]
OK!

Have you made any modifications to the code yet?
Otherwise, I could try to set up the extension tomorrow and note down the steps.

> It might be cool to have a documentation page on how to work on certain parts of the serverpod codebase.

We would love to have this; I'll see if I can produce something when setting it up for this part.
Overall we would like to make it easier to contribute to Serverpod. We will probably do some changes for the projects soon to enable this.

**Comment by vanlooverenkoen** [comment]
I have made changes to see if I can print something and use my own implementation without success. So no, no usefull changes.

Nice!! Let me know if I can test something. I am using this for my own recordings to show something to clients: https://screen.studio/

Might be useful

**Comment by SandPod** [comment]
Tested running the extension today in debug mode, I think that VScode no longer injects the PATH and ENV from the environment it is run in, and therefore, the Serverpod command was not available in the debug environment.

This PR adds a copy for everything that is required from ENV in order to run the extension: https://github.com/serverpod/serverpod/pull/3158

>  I will do some more debugging. because it seems that now my global serverpod_cli installation for other projects is also broken.

After this PR is merged, if Serverpod is working locally, it should work together with the extension.

**Comment by vanlooverenkoen** [comment]
That might explain why it was not available!! Thanks @SandPod 🙏! I have quite a busy 2 weeks, but I will try to fit it in somewhere

**Comment by vanlooverenkoen** [comment]
Alright @SandPod things are slowing down right now. Just finished a full weekend doing a charity event to raise money for cancer treatment. 

Tonight I will start with some opensource stuff again! This one is the first one on this list

**Comment by vanlooverenkoen** [comment]
Still the global installed version is used.

**Comment by vanlooverenkoen** [comment]
I created this function to debug 🙈 please tell me this is not how you actually do it! 😂 

```
void sendNotification(Connection connection, String message) {
  connection.sendNotification(
    'window/showMessage',
    ShowMessageParams(
      message: message,
      type: MessageType.Log,
    ).toJson(),
  );
}
```

But this allows me to work on this issue. I tried to implement the `onRenameRequest` but it doesn't get triggerd 😕

**Comment by SandPod** [comment]
Hey @vanlooverenkoen!

Sorry for the late response here! I was out of the office for a while but am back now.

> Still the global installed version is used!

Unfortunately, the globally installed CLI is the one that is used, the fix was to ensure that the extension could be run in debug. The development experience for the extension isn't the best, unfortunately.

> But from the moment I do a print I get this error:

To further worsen the DevEx no prints are allowed when running the LSP server. All output from the CLI when it is used as LSP is consumed by the extension and needs to align with the protocol.

> I created this function to debug...

I think this is mainly how the development was done. There is also a way to enable verbose communication logging for the LSP server.

If you add this to the `package.json` and run the extension in debug it should be available in outputs under either LSP or serverpod something.

```
 "serverpodLanguageServer.trace.server": {
          "scope": "window",
          "type": "string",
          "enum": [
            "off",
            "messages",
            "verbose"
          ],
          "default": "verbose",
          "description": "Traces the communication between VS Code and the language server."
        }
```

The commit where this debug information was introduced when developing the extension is available here: https://github.com/SandPod/serverpod/blob/837798c8e5a0c948acb97916460d7456fe5ed6c7/tools/serverpod_vscode_extension/package.json

**Comment by vanlooverenkoen** [comment]
Alright! thanks, I might start debugging the [LSP server](https://pub.dev/packages/lsp_server) because the rename is not triggered in the cli itself.

**Comment by Isakdl** [comment]
@vanlooverenkoen stdout is reserved for communication between the LSP server and client, but you can log to stderr, e.g.

```dart
stderr.writeln('My message');
```

If you turn on the message debug log that @SandPod suggestion you can see all the messages sent by both the client and the server. 

One thing you also need to think about is that during the initialization of the LSP server, the server will tell the client what capabilities it has. Only the features that you say you support will be used by the client. See the official protocol documentation for file renames here: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didRename

**Comment by Isakdl** [comment]
>Document renames should be signaled to a server sending a document close notification with the document’s old name followed by an open notification using the document’s new name. Major reason is that besides the name other attributes can change as well like the language that is associated with the document. In addition the new document could not be of interest for the server anymore.

@vanlooverenkoen FYI, according to the protocol documentation you are supposed to get a did open and a did close event for a file rename.

**Comment by vanlooverenkoen** [comment]
Yes but to me it is not yet clear where to get the original and new file names. But I will do some more investigation.
